x/concentrated-liquidity/position.go

```diff
@@ -249,6 +251,81 @@ func (k Keeper) CreateFullRangePosition(ctx sdk.Context, concentratedPool types.
+	// Create a full range (min to max tick) concentrated liquidity position.
+	positionId, amount0, amount1, liquidity, joinTime, err = k.CreateFullRangePosition(ctx, concentratedPool, owner, coins)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Mint cl shares for the position and lock them for the remaining lock duration.
+	// Also sets the position ID to underlying lock ID mapping.
+	concentratedLockId, _, err := k.MintSharesLockAndUpdate(ctx, concentratedPool, positionId, owner, remainingLockDuration, liquidity)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, nil
+}
+
+// CreateFullRangePositionUnlocking creates a full range (min to max tick) concentrated liquidity position for the given pool ID, owner, and coins.
+// This function is strictly used when migrating a balancer position to CL, where the balancer position is currently unlocking.
+// We lock the cl position for whatever the remaining time is from the balancer position and immediately begin unlocking from where it left off.
+func (k Keeper) CreateFullRangePositionUnlocking(ctx sdk.Context, concentratedPool types.ConcentratedPoolExtension, owner sdk.AccAddress, coins sdk.Coins, remainingLockDuration time.Duration) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, concentratedLockID uint64, err error) {
+	// Create a full range (min to max tick) concentrated liquidity position.
+	positionId, amount0, amount1, liquidity, joinTime, err = k.CreateFullRangePosition(ctx, concentratedPool, owner, coins)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Mint cl shares for the position and lock them for the remaining lock duration.
+	// Also sets the position ID to underlying lock ID mapping.
+	concentratedLockId, underlyingLiquidityTokenized, err := k.MintSharesLockAndUpdate(ctx, concentratedPool, positionId, owner, remainingLockDuration, liquidity)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	// Begin unlocking the newly created concentrated lock.
+	concentratedLockID, err = k.lockupKeeper.BeginForceUnlock(ctx, concentratedLockId, underlyingLiquidityTokenized)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, err
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, concentratedLockID, nil
+}
+
+// MintSharesLockAndUpdate mints the shares for the concentrated liquidity position and locks them for the given duration. It also updates the position ID to underlying lock ID mapping.
+func (k Keeper) MintSharesLockAndUpdate(ctx sdk.Context, concentratedPool types.ConcentratedPoolExtension, positionId uint64, owner sdk.AccAddress, remainingLockDuration time.Duration, liquidity sdk.Dec) (concentratedLockID uint64, underlyingLiquidityTokenized sdk.Coins, err error) {
```

Comment:
The only reason we supply pool as param is to get an id,, let's provide id instead

Reply:
https://github.com/osmosis-labs/osmosis/pull/4883/commits/6705164bffaa24fe5db8d185cf446614f122366a